hydra-node/test/Hydra/Chain/Direct/TxSpec.hs

```diff
@@ -141,6 +187,215 @@ spec =
+                        & counterexample ("commit validity: " <> show commitValidity)
+                , let blueprintInputs = blueprintBody ^. inputsTxBodyL
+                      commitInputs = commitTxBody ^. inputsTxBodyL
+                   in property (blueprintInputs `Set.isSubsetOf` commitInputs)
+                        & counterexample ("blueprint inputs: " <> show blueprintInputs)
+                        & counterexample ("commit inputs: " <> show commitInputs)
+                , let blueprintOutputs = toList $ blueprintBody ^. outputsTxBodyL
+                      commitOutputs = toList $ commitTxBody ^. outputsTxBodyL
+                   in property
+                        ( all
+                            (`notElem` blueprintOutputs)
+                            commitOutputs
+                        )
+                        & counterexample ("blueprint outputs: " <> show blueprintOutputs)
+                        & counterexample ("commit outputs: " <> show commitOutputs)
+                , let blueprintSigs = blueprintBody ^. reqSignerHashesTxBodyL
+                      commitSigs = commitTxBody ^. reqSignerHashesTxBodyL
+                   in property (blueprintSigs `Set.isSubsetOf` commitSigs)
+                        & counterexample ("blueprint signatures: " <> show blueprintSigs)
+                        & counterexample ("commit signatures: " <> show commitSigs)
+                , let blueprintRefInputs = blueprintBody ^. referenceInputsTxBodyL
+                      commitRefInputs = commitTxBody ^. referenceInputsTxBodyL
+                   in property (blueprintRefInputs `Set.isSubsetOf` commitRefInputs)
+                        & counterexample ("blueprint reference inputs: " <> show blueprintRefInputs)
+                        & counterexample ("commit reference inputs: " <> show commitRefInputs)
+                ]
+                & cover 1 (spendsFromScript (lookupUTxO, fromLedgerTx blueprintTx)) "blueprint spends script UTxO"
+                & cover 1 (spendsFromPubKey (lookupUTxO, fromLedgerTx blueprintTx)) "blueprint spends pub key UTxO"
+                & cover 1 (not . null $ blueprintTx ^. bodyTxL . referenceInputsTxBodyL) "blueprint has reference input"
+
+getAuxMetadata :: AlonzoTxAuxData LedgerEra -> Map Word64 Metadatum
+getAuxMetadata (AlonzoTxAuxData metadata _ _) = metadata
+
+addRequiredSignatures :: [VerificationKey PaymentKey] -> Tx -> Tx
+addRequiredSignatures vks tx =
+  fromLedgerTx $
+    toLedgerTx tx
+      & bodyTxL . reqSignerHashesTxBodyL .~ Set.fromList (toLedgerKeyHash . verificationKeyHash <$> vks)
+
+genBlueprintTxWithUTxO :: Gen (UTxO, Tx)
+genBlueprintTxWithUTxO =
+  fmap (second unsafeBuildTransaction) $
+    spendingPubKeyOutput (mempty, emptyTxBody)
+      >>= spendSomeScriptInputs
+      >>= addSomeReferenceInputs
+      >>= addValidityRange
+      >>= addRandomMetadata
+      >>= removeRandomInputs
```

Comment:
Why is `removeRandomInputs` here?

As we are building up a random blueprint transaction from scratch it does not make much sense to remove inputs?

Reply:
We start from the empty UTxO and then incrementally add some UTxO and more details to tx body. I wanted to remove some tx input arbitrarily and keep the UTxO the same so that we exercise what happens when user submits a UTxO and doesn't want all of it's inputs committed.